File: Csharp/Leetcode/0092_ReverseLinkedList2/ReverseBetween.cs
using System;
using Leetcode.Utils;
namespace Leetcode.Solutions {
    public class ReverseBetween {
        public ListNode Solution(ListNode head, int m, int n) {
            if (m < 1 || n < m || head == null) {
                return null;
            }


            var i = 1;
            ListNode portBegin = null;
            ListNode portEnd = null;
            var curNode = head;
            ListNode preNode = null;
            ListNode tempNode = null;
            ListNode startNode = null;

            while (curNode != null) {
                if (m != 1 && i + 1 == m) {
                    portBegin = curNode;   
                }

                if (i == m) {
                    startNode = curNode;
                }

                if (i == n) {
                    portEnd = curNode.next;
                }

                if (i >= m && i <= n) {
                    tempNode = curNode;
                    curNode = curNode.next;
                    tempNode.next = preNode;
                    preNode = tempNode;
                } else {
                    curNode = curNode.next;
                }

                i++;
            }

            if (i <= n) {
                return null;
            }

            startNode.next = portEnd;
            if (portBegin != null) {
                portBegin.next = tempNode;
                return head;
            }
            return tempNode;
        }

        private ListNode Reverse(ListNode head) {

            if (head == null) {
                return null;
            }

            var curNode = head;
            ListNode preNode = null;
            ListNode tempNode = null;

            while (curNode != null) {
                tempNode = curNode;
                curNode = curNode.next;
                tempNode.next = preNode;
                preNode = tempNode;
            }

            return tempNode;
        }
    }
}

File: Python/0092_ReverseLinkedList2/reverseBetween.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """
        if m < 1 or m > n or not head:
            return None
        
        cur, pre = head, None

        for _ in range(m - 1):
            pre = cur
            cur = cur.next
        
        endFirst, startSecond = pre, cur        
        pre = cur
        cur = cur.next

        for _ in range(n - m):
            next = cur.next
            cur.next = pre
            pre = cur
            cur = next

        startSecond.next = cur
        if m == 1:
            return pre

        endFirst.next = pre        
        return head